src/Suite.ts

```diff
@@ -1,52 +1,99 @@
 /* eslint-env node, mocha */
-import assert from 'assert';
-import each from 'lodash/each';
+import defer from 'lodash/defer';
 import groupBy from 'lodash/groupBy';
-import options from './Options';
+import { createEmitter, Emitter } from './Emitter';
+import { Options } from './Options';
 import * as ProcessedStory from './ProcessedStory';
 import * as Result from './Result';
 import * as TestBrowser from './TestBrowser';
 
 /**
- * Find Storybook stories and generate a test for each one.
+ * Mapping of event names to handlers for the test suite.
  */
-async function writeTests() {
-  const testBrowser = await TestBrowser.create(options);
-  const page = await TestBrowser.createPage(testBrowser, options);
-  const stories = await TestBrowser.getStories(page);
-  const storiesByComponent = groupBy(stories, 'componentTitle');
-
-  describe(`[${options.browser}] accessibility`, function () {
-    after(async function () {
+export type SuiteEvents = {
+  suiteStart: (browser: string) => void;
+  componentStart: (componentName: string) => void;
+  componentSkip: (componentName: string) => void;
+  storyStart: (storyName: string) => void;
+  storyPass: (storyName: string, result: Result.Result, elapsedTime: number) => void;
+  storyFail: (storyName: string, result: Result.Result, elapsedTime: number) => void;
+  storySkip: (storyName: string) => void;
+  suiteFinish: (numPass: number, numFail: number, numSkip: number, elapsedTime: number) => void;
+}
+
+/**
+ * Event emitter representing a run of the test suite.
+ */
+export type SuiteEmitter = Emitter<SuiteEvents>;
+
+/**
+ * Find Storybook stories and run Axe on each one. Returns an event emitter that emits events when
+ * components and stories are processed.
+ */
+export function run(options: Options): SuiteEmitter {
+  const emitter = createEmitter<SuiteEvents>();
+
+  defer(async () => {
```

Comment:
does `defer` do the same as `delay: true` in Mocha? i.e. making sure the tests are only run after they're all defined?

Reply:
Sort of.

`delay: true` was delaying the start of the Mocha test suite until we'd finished defining the tests (which required some async operations).

`defer` here is just delaying the start of the test suite to the back of the event loop, which lets us return the event emitter before anything starts happening.

Could have achieved the same thing with

```js
setTimeout(async () => {
  // Stuff
}, 0);
```

Reply:
ohhh gotcha, thanks for explaining!